naga/src/back/spv/block.rs

```diff
@@ -2077,14 +2096,60 @@ impl<'w> BlockContext<'w> {
-                    self.write_block(
+                    let merge_used = self.write_block(
                         scope_id,
                         block_statements,
                         BlockExit::Branch { target: merge_id },
                         loop_context,
                         debug_info,
                     )?;
 
+                    if !merge_used {
+                        // Avoid generating unreachable SPIR-V blocks, because they might
+                        // try to jump back into loops in violation of SPIR-V's structured
+                        // control flow rules.
+                        //
+                        // For example, the WGSL front end generates code like this:
+                        //
+                        //     Loop {
+                        //       body: Block {
+                        //         If i < 4 {
+                        //           accept: Block { }
+                        //           reject: Block { Break }
+                        //         }
+                        //         Block { Break }
+                        //       }
+                        //       continuing: Block {
+                        //         i += 1;
+                        //       }
+                        //     }
+                        //
+                        // Suppose this call to `write_block` is handling the `Loop` body.
+                        // Thus, our `exit` argument is a `BlockExit::Branch`, jumping to
+                        // the `Loop`'s continuing block.
+                        //
+                        // When this call reaches the `Block { Break }`, the call to
+                        // `write_block` above will handle it. However, that recursive
+                        // call will return without ever using the `BlockExit` we pass to
+                        // it, since the block has a `Break`.
+                        //
+                        // If we ignore `merge_used` and naively start a new block, this
+                        // new block will be unreachable. However, when we finish the
+                        // current Naga `Block`, we will terminate this new unreachable
+                        // block according to our `exit` value, and try to branch to the
+                        // loop's continuing block.
+                        //
+                        // The result is that we will introduce an unreachable SPIR-V
+                        // block that branches into the loop construct.
+                        //
+                        // It seems to me that if a block is unreachable, its behavior
+                        // shouldn't affect the validity of the SPIR-V module. The newer
+                        // versions of the SPIR-V specification talk about "structural
+                        // reachability" that I would have expected to address this. But
+                        // the SPIR-V validator still complains.
```

Comment:
style: Please don't use singular first-person in a shared codebase; it doesn't really have meaning in code like this. 😅